### PR TITLE
Implemented C# 8 Nullable Reference Type standard.

### PR DESCRIPTION
**Pull request by pmaytak**
Issue #15 
Original branch #183

- Implemented C# 8 Nullable Reference Type standard.
- Includes additional null checks for public method parameters.
- Few cases where I had to refactor a little to prevent null exceptions.
- Misc warning and comment fixes.

**Note 1:** 
Created a separate branch because rebasing #183 had some conflicts. Work from #183 is included here.

**Note 2:**
Analyzers had some funky behavior. Some code didn't have squiggly underlines. Error list had some weird warning types. Perhaps something to do with preview .NET build?
image

References:
[Nullable reference types](https://docs.microsoft.com/en-us/dotnet/csharp/nullable-references)
[Reserved attributes contribute to the compiler's null state static analysis](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/attributes/nullable-analysis)
[Try out Nullable Reference Types](https://devblogs.microsoft.com/dotnet/try-out-nullable-reference-types/)
[Update libraries to use nullable reference types and communicate nullable rules to callers](https://docs.microsoft.com/en-us/dotnet/csharp/nullable-migration-strategies)

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -527,20 +524,23 @@ dotnet_diagnostic.SA1614.severity = none
...
 dotnet_diagnostic.SA1649.severity = none
 #########################################################
 #########################################################
 
+# SA1011: Closing square brackets should be spaced correctly
+dotnet_diagnostic.SA1011.severity = none
+
+# SA1009: Closing parenthesis should be spaced correctly
+dotnet_diagnostic.SA1009.severity = none
+
```
Comment (pmaytak) [inline .editorconfig:543]: Current version of StyleCop Analyzers has bugs related to `?` and `!` operators which are fixed in version 1.2.0 but it's still beta. So commented these rules out to reduce noise.

File: src/Microsoft.Identity.Web/AuthorityHelpers.cs [hunk 2]
```diff
@@ -10,8 +10,8 @@ internal static class AuthorityHelpers
     {
         internal static string BuildAuthority(MicrosoftIdentityOptions options)
         {
-            var baseUri = new Uri(options.Instance);
-            var pathBase = baseUri.PathAndQuery.TrimEnd('/');
+            Uri baseUri = new Uri(options.Instance);
```
Comment (pmaytak) [inline src/Microsoft.Identity.Web/AuthorityHelpers.cs:13]: By default, `var` becomes the nullable reference type (`Uri?` in this case). So in some cases I used explicit type instead.

File: src/Microsoft.Identity.Web/CertificateManagement/DefaultCertificateLoader.cs [hunk 3]
```diff
@@ -27,19 +27,19 @@ public void LoadIfNeeded(CertificateDescription certificateDescription)
                 switch (certificateDescription.SourceType)
                 {
                     case CertificateSource.KeyVault:
-                        certificateDescription.Certificate = LoadFromKeyVault(certificateDescription.Container, certificateDescription.ReferenceOrValue);
+                        certificateDescription.Certificate = LoadFromKeyVault(certificateDescription.Container!, certificateDescription.ReferenceOrValue!);
```
Comment (pmaytak) [inline src/Microsoft.Identity.Web/CertificateManagement/DefaultCertificateLoader.cs:30]: `!` overrides the conversion from nullable to non-nullable type, which otherwise would throw an exception. In this case, the properties shouldn't be null.

File: src/Microsoft.Identity.Web/CertificateManagement/DefaultCertificateLoader.cs [hunk 4]
```diff
@@ -174,11 +174,7 @@ private static void ParseStoreLocationAndName(string storeDescription, ref Store
         /// <summary>
         /// Find a certificate by criteria.
         /// </summary>
-        /// <param name="x509Store"></param>
-        /// <param name="identifierCriterium"></param>
-        /// <param name="certificateIdentifier"></param>
-        /// <returns></returns>
-        private static X509Certificate2 FindCertificateByCriterium(
+        private static X509Certificate2? FindCertificateByCriterium(
```
Comment (pmaytak) [inline src/Microsoft.Identity.Web/CertificateManagement/DefaultCertificateLoader.cs:177]: Wondering if instead of returning `null` we throw an exception in this method, if the certificate is not found. Other load methods will throw exceptions if the certificate can't be created.

File: src/Microsoft.Identity.Web/CertificateManagement/DefaultCertificateLoader.cs [hunk 5]
```diff
@@ -203,7 +199,7 @@ private static void ParseStoreLocationAndName(string storeDescription, ref Store
             DefaultCertificateLoader defaultCertificateLoader = new DefaultCertificateLoader();
             CertificateDescription certDescription = certificateDescription.First();
             defaultCertificateLoader.LoadIfNeeded(certDescription);
-            return certDescription?.Certificate;
+            return certDescription.Certificate!;
```
Comment (pmaytak) [inline src/Microsoft.Identity.Web/CertificateManagement/DefaultCertificateLoader.cs:202]: `Certificate` here, I assume, should never be null. But if it's not found in the cert store, it will be null. I think this will result in passing null to MSAL and to Wilson.

File: src/Microsoft.Identity.Web/InstanceDiscovery/IssuerConfigurationRetriever.cs [hunk 6]
```diff
@@ -40,7 +42,7 @@ public async Task<IssuerMetadata> GetConfigurationAsync(string address, IDocumen
             };
 
             string doc = await retriever.GetDocumentAsync(address, cancel).ConfigureAwait(false);
-            return JsonSerializer.Deserialize<IssuerMetadata>(doc, options);
+            return JsonSerializer.Deserialize<IssuerMetadata>(doc, options)!;
```
Comment (pmaytak) [inline src/Microsoft.Identity.Web/InstanceDiscovery/IssuerConfigurationRetriever.cs:45]: This case is weird - the compiler says `Deserialize` can return null, but the docs don't mention it, say it just throws exceptions.

File: src/Microsoft.Identity.Web/MicrosoftIdentityOptions.cs [hunk 7]
```diff
@@ -16,17 +16,17 @@ public class MicrosoftIdentityOptions : OpenIdConnectOptions
         /// <summary>
         /// Gets or sets the Azure Active Directory instance, e.g. "https://login.microsoftonline.com".
         /// </summary>
-        public string Instance { get; set; }
+        public string Instance { get; set; } = null!;
```
Comment (pmaytak) [inline src/Microsoft.Identity.Web/MicrosoftIdentityOptions.cs:19]: This is a workaround for properties that are really non-nullable, except in once instance when the Data Transfer Objects are not yet initialized.

File: tests/.editorconfig [hunk 8]
```diff
@@ -5,6 +5,9 @@
 # Code files
 [*.{cs,vb}]
 
+# SA0001: XML comment analysis disabled
+dotnet_diagnostic.SA0001.severity = none
```
Comment (pmaytak) [inline tests/.editorconfig:9]: This removes 4 warnings from the build pipeline related to disabled comment analysis.

File: src/Microsoft.Identity.Web/MicrosoftIdentityOptions.cs [hunk 9]
```diff
@@ -39,7 +39,7 @@ public class MicrosoftIdentityOptions : OpenIdConnectOptions
         /// If you don't specify the redirect URI, the redirect URI will be computed from the URL on which the app is
         /// deployed and the CallbackPath.
         /// </summary>
-        public string RedirectUri { get; set; }
+        public string? RedirectUri { get; set; }
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/MicrosoftIdentityOptions.cs:42]: > public string? RedirectUri { get; set; } [](start = 7, length = 41)

need to update w/master, these have been removed.

File: src/Microsoft.Identity.Web/MicrosoftIdentityOptions.cs [hunk 10]
```diff
@@ -70,25 +70,25 @@ public class MicrosoftIdentityOptions : OpenIdConnectOptions
...
 
         /// <summary>
         /// Gets the default user flow (which is signUpsignIn).
         /// </summary>
-        public string DefaultUserFlow => SignUpSignInPolicyId;
+        public string? DefaultUserFlow => SignUpSignInPolicyId;
 
         /// <summary>
-        /// Is considered B2C if the attribute SignUpSignInPolicyId is defined.
+        /// Gets a value indicating whether app is considered B2C if the attribute SignUpSignInPolicyId is defined.
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/MicrosoftIdentityOptions.cs:91]: nit: the previous comment is clearer imho

File: src/Microsoft.Identity.Web/TokenAcquisition.cs [hunk 11]
```diff
@@ -539,7 +535,7 @@ public void ReplyForbiddenWithWwwAuthenticateHeader(IEnumerable<string> scopes,
                 headers.Remove(HeaderNames.WWWAuthenticate);
             }
 
-            headers.Add(HeaderNames.WWWAuthenticate, v);
+            headers.Add(HeaderNames.WWWAuthenticate, new StringValues($"Bearer {parameterString}"));
         }
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/TokenAcquisition.cs:539]: this has been changed as well in latest master.

File: src/Microsoft.Identity.Web/TokenAcquisition.cs [hunk 12]
```diff
@@ -502,8 +498,8 @@ private async Task<IConfidentialClientApplication> BuildConfidentialClientApplic
         /// the client can trigger an interaction with the user so that the user consents to more scopes.
         /// </summary>
         /// <param name="scopes">Scopes to consent to.</param>
-        /// <param name="msalServiceException"><see cref="MsalUiRequiredException"/> triggering the challenge.</param>
-        public void ReplyForbiddenWithWwwAuthenticateHeader(IEnumerable<string> scopes, MsalUiRequiredException msalServiceException)
+        /// <param name="msalServiceException">The <see cref="MsalUiRequiredException"/> that triggered the challenge.</param>
+        public async Task ReplyForbiddenWithWwwAuthenticateHeader(IEnumerable<string> scopes, MsalUiRequiredException msalServiceException)
         {
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/TokenAcquisition.cs:503]: method name should end in `Async` if it's async now.

File: src/Microsoft.Identity.Web/WebAppCallsWebApiAuthenticationBuilderExtensions.cs [hunk 13]
```diff
@@ -160,7 +174,7 @@ public static class WebAppCallsWebApiAuthenticationBuilderExtensions
                 options.Events.OnRedirectToIdentityProviderForSignOut = async context =>
                 {
                     // Remove the account from MSAL.NET token cache
-                    var tokenAcquisition = context.HttpContext.RequestServices.GetRequiredService<ITokenAcquisition>() as ITokenAcquisitionInternal;
+                    var tokenAcquisition = (ITokenAcquisitionInternal)context.HttpContext.RequestServices.GetRequiredService<ITokenAcquisition>();
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/WebAppCallsWebApiAuthenticationBuilderExtensions.cs:177]: is this part of a fix for [240](https://github.com/AzureAD/microsoft-identity-web/issues/240)? if so, can it go in a separate PR?

File: src/Microsoft.Identity.Web/CertificateManagement/DefaultCertificateLoader.cs (hunk 5 above)
Comment (jennyf19) [reply src/Microsoft.Identity.Web/CertificateManagement/DefaultCertificateLoader.cs:202]: MSAL will handle the cert being null and throw an `ArgumentNullException`, we catch the exception

---
In reply to: [444031657](https://github.com/AzureAD/microsoft-identity-web/pull/231#discussion_r444031657) [](ancestors = 444031657)

File: src/Microsoft.Identity.Web/CertificateManagement/DefaultCertificateLoader.cs (hunk 4 above)
Comment (jennyf19) [reply src/Microsoft.Identity.Web/CertificateManagement/DefaultCertificateLoader.cs:177]: the keyvault one will throw an exception, and if parsing the location fails, we throw an exception, otherwise we are returning the cert or null on all the other load methods.

---
In reply to: [444030354](https://github.com/AzureAD/microsoft-identity-web/pull/231#discussion_r444030354) [](ancestors = 444030354)

File: src/Microsoft.Identity.Web/InstanceDiscovery/IssuerConfigurationRetriever.cs (hunk 6 above)
Comment (jennyf19) [reply src/Microsoft.Identity.Web/InstanceDiscovery/IssuerConfigurationRetriever.cs:45]: want to make a comment here then for future reference?

---
In reply to: [444034173](https://github.com/AzureAD/microsoft-identity-web/pull/231#discussion_r444034173) [](ancestors = 444034173)

File: src/Microsoft.Identity.Web/Resource/JwtBearerMiddlewareDiagnostics.cs [hunk 14]
```diff
@@ -52,10 +52,7 @@ public JwtBearerMiddlewareDiagnostics(ILogger<JwtBearerMiddlewareDiagnostics> lo
         /// <param name="events">Events to subscribe to.</param>
         public JwtBearerEvents Subscribe(JwtBearerEvents events)
         {
-            if (events == null)
-            {
-                events = new JwtBearerEvents();
-            }
+            events ??= new JwtBearerEvents();
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/Resource/JwtBearerMiddlewareDiagnostics.cs:55]: nit: space?

File: src/Microsoft.Identity.Web/Resource/OpenIdConnectMiddlewareDiagnostics.cs [hunk 15]
```diff
@@ -80,6 +80,8 @@ public OpenIdConnectMiddlewareDiagnostics(ILogger<OpenIdConnectMiddlewareDiagnos
         /// <param name="events">Events to subscribe to.</param>
         public void Subscribe(OpenIdConnectEvents events)
         {
+            events ??= new OpenIdConnectEvents();
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/Resource/OpenIdConnectMiddlewareDiagnostics.cs:83]: nit: space?

File: src/Microsoft.Identity.Web/Resource/RegisterValidAudience.cs [hunk 16]
```diff
@@ -61,7 +61,12 @@ internal class RegisterValidAudience
             SecurityToken securityToken,
             TokenValidationParameters validationParameters)
         {
-            JwtSecurityToken token = securityToken as JwtSecurityToken;
+            JwtSecurityToken? token = securityToken as JwtSecurityToken;
+            if (token == null)
+            {
+                throw new SecurityTokenValidationException("Token is not JWT token.");
+            }
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/Resource/RegisterValidAudience.cs:68]: is that the right error message? when would the casting fail in this case and result in a null token value? the message should have some customer value, so how can they resolve or understand the issue to fix it.

**Comment by jennyf19** [comment]
some comments. it's hard to review w/so many xml and spelling changes. in the future, can you do those in a separate PR, especially when you have so many? it tends to create noise and slows down the review. thank you.

File: src/Microsoft.Identity.Web/WebAppCallsWebApiAuthenticationBuilderExtensions.cs (hunk 13 above)
Comment (pmaytak) [reply src/Microsoft.Identity.Web/WebAppCallsWebApiAuthenticationBuilderExtensions.cs:177]: No, this is just to make it non-nullable. #240 will change this line anyway.

**Comment by pmaytak** [comment]
> some comments. it's hard to review w/so many xml and spelling changes. in the future, can you do those in a separate PR, especially when you have so many? it tends to create noise and slows down the review. thank you.

@jennyf19 Will do.

File: .editorconfig (hunk 1 above)
Comment (jmprieur) [reply .editorconfig:543]: why not leverage the beta? we are beta ourselves?

File: src/Microsoft.Identity.Web/AuthorizeForScopesAttribute.cs [hunk 17]
```diff
@@ -32,38 +32,37 @@ public class AuthorizeForScopesAttribute : ExceptionFilterAttribute
...
+        public string? ScopeKeySection { get; set; }
 
         /// <summary>
-        /// Handles the MsalUiRequiredException.
+        /// Handles the <see cref="MsalUiRequiredException"/>.
         /// </summary>
         /// <param name="context">Context provided by ASP.NET Core.</param>
         public override void OnException(ExceptionContext context)
         {
-            // Do not re-use the attribute param Scopes. For more info: https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/issues/273
```
Comment (jmprieur) [inline src/Microsoft.Identity.Web/AuthorizeForScopesAttribute.cs:48]: why remove this comment?

File: src/Microsoft.Identity.Web/ITokenAcquisition.cs [hunk 18]
```diff
@@ -42,7 +42,7 @@ public interface ITokenAcquisition
         /// </summary>
         /// <param name="scopes">Scopes to consent to.</param>
         /// <param name="msalSeviceException"><see cref="MsalUiRequiredException"/> triggering the challenge.</param>
-        void ReplyForbiddenWithWwwAuthenticateHeader(
+        Task ReplyForbiddenWithWwwAuthenticateHeader(
```
Comment (jmprieur) [inline src/Microsoft.Identity.Web/ITokenAcquisition.cs:45]: postpone with Async?

File: .editorconfig (hunk 1 above)
Comment (pmaytak) [reply .editorconfig:543]: I can propose that change in a separate PR.

File: src/Microsoft.Identity.Web/AuthorizeForScopesAttribute.cs (hunk 17 above)
Comment (jennyf19) [reply src/Microsoft.Identity.Web/AuthorizeForScopesAttribute.cs:48]: it's further down.

File: src/Microsoft.Identity.Web/CertificateManagement/DefaultCertificateLoader.cs (hunk 4 above)
Comment (pmaytak) [reply src/Microsoft.Identity.Web/CertificateManagement/DefaultCertificateLoader.cs:177]: If the certificate is not found, FromDisk, FromBase64, and FromKeyVault throw an exception; two FromCertStore methods return null. So my thought was to make the behavior consistent. Perhaps log that cert was not found and return null? This might benefit the certificate rotation logic (when implemented) when multiple certs are specified and we want to try until valid one is found.

File: src/Microsoft.Identity.Web/CertificateManagement/DefaultCertificateLoader.cs (hunk 5 above)
Comment (pmaytak) [reply src/Microsoft.Identity.Web/CertificateManagement/DefaultCertificateLoader.cs:202]: - We do get an uncaught NPE when trying to set token decryption cert.
https://github.com/AzureAD/microsoft-identity-web/blob/6fe02dbb312bed2d7351ff9035a6e717cafc7574/src/Microsoft.Identity.Web/WebApiAuthenticationBuilderExtensions.cs#L117-L118

- Yes, we do catch the NPE from MSAL. But the exception is kind of general. Also, FromBase64 gives `Cannot find the requested object.`; FromDisk gives `The system cannot find the file specified.` Similar to my other comment, maybe we can update this to give better messaging and user experience when we implement cert rotation stuff.

File: src/Microsoft.Identity.Web/Resource/JwtBearerMiddlewareDiagnostics.cs (hunk 14 above)
Comment (pmaytak) [reply src/Microsoft.Identity.Web/Resource/JwtBearerMiddlewareDiagnostics.cs:55]: Where? Looks right.

File: src/Microsoft.Identity.Web/Resource/RegisterValidAudience.cs (hunk 16 above)
Comment (pmaytak) [reply src/Microsoft.Identity.Web/Resource/RegisterValidAudience.cs:68]: I am not sure if it's always `JwtSecurityToken`. Is it? If it is, I can just explicitly cast it.

File: src/Microsoft.Identity.Web/AuthorizeForScopesAttribute.cs (hunk 17 above)
Comment (pmaytak) [reply src/Microsoft.Identity.Web/AuthorizeForScopesAttribute.cs:48]: It was moved few lines down.
https://github.com/AzureAD/microsoft-identity-web/blob/6fe02dbb312bed2d7351ff9035a6e717cafc7574/src/Microsoft.Identity.Web/AuthorizeForScopesAttribute.cs#L63-L64

File: src/Microsoft.Identity.Web/MicrosoftIdentityOptions.cs (hunk 10 above)
Comment (pmaytak) [reply src/Microsoft.Identity.Web/MicrosoftIdentityOptions.cs]: Yea, I agree. Reverted it. I'm not a huge fan of that analyzer suggestion myself.

File: src/Microsoft.Identity.Web/InstanceDiscovery/IssuerConfigurationRetriever.cs (hunk 6 above)
Comment (pmaytak) [reply src/Microsoft.Identity.Web/InstanceDiscovery/IssuerConfigurationRetriever.cs]: Added.

File: src/Microsoft.Identity.Web/TokenAcquisition.cs (hunk 12 above)
Comment (pmaytak) [reply src/Microsoft.Identity.Web/TokenAcquisition.cs:503]: Good catch. Updated.

File: src/Microsoft.Identity.Web/ITokenAcquisition.cs (hunk 18 above)
Comment (pmaytak) [reply src/Microsoft.Identity.Web/ITokenAcquisition.cs]: Good catch. Updated.

File: src/Microsoft.Identity.Web/MicrosoftIdentityOptions.cs (hunk 9 above)
Comment (pmaytak) [reply src/Microsoft.Identity.Web/MicrosoftIdentityOptions.cs]: Rebased.

File: src/Microsoft.Identity.Web/TokenAcquisition.cs (hunk 11 above)
Comment (pmaytak) [reply src/Microsoft.Identity.Web/TokenAcquisition.cs:541]: Rebased.

File: src/Microsoft.Identity.Web/CertificateManagement/DefaultCertificateLoader.cs (hunk 5 above)
Comment (jennyf19) [reply src/Microsoft.Identity.Web/CertificateManagement/DefaultCertificateLoader.cs]: That's fine, but maybe something for a separate PR w/a new issue.

File: src/Microsoft.Identity.Web/MicrosoftIdentityOptions.cs (hunk 10 above)
Comment (jennyf19) [reply src/Microsoft.Identity.Web/MicrosoftIdentityOptions.cs]: The analyzer suggests that?

File: src/Microsoft.Identity.Web/Resource/RegisterValidAudience.cs (hunk 16 above)
Comment (jennyf19) [reply src/Microsoft.Identity.Web/Resource/RegisterValidAudience.cs:68]: might be something to look into. the error message is misleading.
